File: trendmicro/FunnyCalculator/FunnyCalculator.java
import java.io.IOException;
import java.io.PushbackInputStream;

public class Task {
	private static enum TOKEN_TYPE {
		L_BRACKET,
		R_BRACKET,
		ADD,
		MINUS,
		MUL,
		DIV,
		NUMBER,
		NIL
	}

	static int next_number;

	static PushbackInputStream pis = new PushbackInputStream(System.in);

	private static TOKEN_TYPE scanner() throws IOException {
		int c;
		while (true) {
			c = pis.read();
			if (Character.isWhitespace(c)) {
				continue;
			}
			break;
		};
		if (c == '+') {
			return TOKEN_TYPE.ADD;
		}
		else if (c == '-') {
			return TOKEN_TYPE.MINUS;
		}
		else if (c == '*') {
			return TOKEN_TYPE.MUL;
		}
		else if (c == '/') {
			return TOKEN_TYPE.DIV;
		}
		else if (Character.isLetter(c)) {
			next_number = Character.toUpperCase(c) - 'A';
			int x = pis.read();
			while(Character.isLetter(x) || Character.isWhitespace(x)){
				if (Character.isWhitespace(x)) {
					x = pis.read();
					continue;
				}
				next_number *= 10;
				next_number += Character.toUpperCase(x) - 'A';
				x = pis.read();
			};
			pis.unread(x);
			return TOKEN_TYPE.NUMBER;
		}
		else if (c == '(') {
			return TOKEN_TYPE.L_BRACKET;
		}
		else if (c == ')') {
			return TOKEN_TYPE.R_BRACKET;
		}
		return TOKEN_TYPE.NIL;
	}

	private static TOKEN_TYPE next_token = TOKEN_TYPE.NIL;

	private static TOKEN_TYPE match(TOKEN_TYPE got_token) throws IOException {
		if (got_token == next_token) {
			next_token = scanner();
		}
		else {
			;
		}
		return next_token;
	}

	private static int expr() throws IOException {
		int acc;
		acc = term();
		while (next_token == TOKEN_TYPE.ADD || next_token == TOKEN_TYPE.MINUS) {
			if (next_token == TOKEN_TYPE.ADD) {
				match(TOKEN_TYPE.ADD);
				acc += term();
			}
			else if (next_token == TOKEN_TYPE.MINUS) {
				match(TOKEN_TYPE.MINUS);
				acc -= term();
			};
		};
		return acc;
	}

	private static int term() throws IOException {
		int acc;
		acc = factor();
		while (next_token == TOKEN_TYPE.MUL || next_token == TOKEN_TYPE.DIV) {
			if (next_token == TOKEN_TYPE.MUL) {
				match(TOKEN_TYPE.MUL);
				acc *= factor();
			}
			else if (next_token == TOKEN_TYPE.DIV) {
				match(TOKEN_TYPE.DIV);
				acc /= factor();
			};
		};
		return acc;
	}

	private static int factor() throws IOException {
		int r = 0;
		if (next_token == TOKEN_TYPE.NUMBER) {
			r = next_number;
			match(TOKEN_TYPE.NUMBER);
		}
		else if (next_token == TOKEN_TYPE.L_BRACKET) {
			match(TOKEN_TYPE.L_BRACKET);
			r = expr();
			match(TOKEN_TYPE.R_BRACKET);
		}
		return r;
	}

	public static void main(String[] args) throws IOException {
		match(TOKEN_TYPE.NIL);
		System.out.println(String.format("%d", expr()));
	}
}

File: trendmicro/FunnyCalculator/FunnyCalculator.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

typedef enum {
    L_BRACKET,
    R_BRACKET,
    ADD,
    MINUS,
    MUL,
    DIV,
    NUMBER,
    NIL
} TOKEN_TYPE;

static int next_number;
TOKEN_TYPE scanner (void){
    char c;
    while(1){
        c = fgetc(stdin);
        if (isblank(c)) {
            continue;
        }
        break;
    };
    if (c == '+') {
        return ADD;
    }
    else if (c == '-') {
        return MINUS;
    }
    else if (c == '*') {
        return MUL;
    }
    else if (c == '/') {
        return DIV;
    }
    else if (isalpha(c)) {
        next_number = (int) toupper(c) - 65;
        char x = fgetc(stdin);
        while(isalpha(x) || isblank(x)){
            if (isblank(x)) {
            x = fgetc(stdin);
                continue;
            }
            next_number *= 10;
            next_number += (int) toupper(x) - 65;
            x = fgetc(stdin);
        };
        ungetc(x, stdin);
        return NUMBER;
    }
    else if (c == '(') {
        return L_BRACKET;
    }
    else if (c == ')') {
        return R_BRACKET;
    }
    return NIL;
};

static TOKEN_TYPE next_token = NIL;
TOKEN_TYPE match(TOKEN_TYPE got_token){
    if (got_token == next_token) {
        next_token = scanner();
    }
    else {
        ;
    }
    return next_token;
}

/*
 * <expr> -> <term> { + - <term> }
 * <term> -> <factor> { * / <factor> }
 * <factor> -> N | ( <expr> )
 */
int expr(void);
int term(void);
int factor(void);

int expr(void) {
    int acc;
    acc = term();
    while (next_token == ADD || next_token == MINUS) {
        if (next_token == ADD) {
            match(ADD);
            acc += term();
        }
        else if (next_token == MINUS) {
            match(MINUS);
            acc -= term();
        };
    };
    return acc;
};

int term(void) {
    int acc;
    acc = factor();
    while (next_token == MUL || next_token == DIV) {
        if (next_token == MUL) {
            match(MUL);
            acc *= factor();
        }
        else if (next_token == DIV) {
            match(DIV);
            acc /= factor();
        };
    };
    return acc;
};

int factor(void) {
    int r;
    if (next_token == NUMBER) {
        r = next_number;
        match(NUMBER);
    }
    else if (next_token == L_BRACKET) {
        match(L_BRACKET);
        r = expr();
        match(R_BRACKET);
    }
    return r;
};

int main(void){
    match(NIL);
    printf("%d\n", expr());
    return 0;
};
